homeassistant/components/deluge/switch.py

```diff
@@ -81,35 +64,38 @@ def is_on(self):
     @property
     def available(self):
         """Return true if device is available."""
-        return self._available
+        return self.client.api.available
 
     def turn_on(self, **kwargs):
-        """Turn the device on."""
-        torrent_ids = self.deluge_client.call("core.get_session_state")
-        self.deluge_client.call("core.resume_torrent", torrent_ids)
+        """Start all torrents."""
+        _LOGGER.debug("Starting all torrents")
+        self.client.api.start_torrents()
+        self.client.api.update()
 
     def turn_off(self, **kwargs):
-        """Turn the device off."""
-        torrent_ids = self.deluge_client.call("core.get_session_state")
-        self.deluge_client.call("core.pause_torrent", torrent_ids)
+        """Stop all torrents."""
+        _LOGGER.debug("Stoping all torrents")
+        self.client.api.stop_torrents()
+        self.client.api.update()
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        self.unsub_dispatcher = async_dispatcher_connect(
+            self.hass, self.client.api.signal_update, self._schedule_immediate_update,
+        )
+
+    @callback
+    def _schedule_immediate_update(self):
+        self.async_schedule_update_ha_state(True)
 
     def update(self):
         """Get the latest data from deluge and updates the state."""
-        from deluge_client import FailedToReconnectException
-
-        try:
-            torrent_list = self.deluge_client.call(
-                "core.get_torrents_status", {}, ["paused"]
-            )
-            self._available = True
-        except FailedToReconnectException:
-            _LOGGER.error("Connection to Deluge Daemon Lost")
-            self._available = False
-            return
-        for torrent in torrent_list.values():
-            item = torrent.popitem()
-            if not item[1]:
-                self._state = STATE_ON
-                return
-
-        self._state = STATE_OFF
+        if self.client.api.get_active_torrents_count() > 0:
+            self._state = STATE_ON
```

Comment:
It's actually weird that this is a switch. If you have 0 torrents active, you cannot turn it on ?

Reply:
Turning ON actually means start all torrents. The state of the switch reflects that there are active torrents (either downloading or seeding). In case of 0 this means that we don't have any active torrents which is the same as pausing all torrents in which case the switch state is OFF.